bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/discovery/BridgeDiscoveryParticipant.java

```diff
@@ -0,0 +1,152 @@
+    }
+
+    @Override
+    public @Nullable DiscoveryResult createResult(ServiceInfo serviceInfo) {
+        logger.trace("Bridge Discovery started for {}", serviceInfo);
+
+        @Nullable
+        final ThingUID uid = getThingUID(serviceInfo);
+        if (uid == null) {
+            return null;
+        }
+
+        logger.trace("Discovered Bosch Smart Home Controller at {}", bridgeInformation.shcIpAddress);
+
+        return DiscoveryResultBuilder.create(uid)
+                .withLabel("Bosch Smart Home Controller (" + bridgeInformation.shcIpAddress + ")")
+                .withProperty("ipAddress", bridgeInformation.shcIpAddress)
+                .withProperty("shcGeneration", bridgeInformation.shcGeneration)
+                .withProperty("apiVersions", bridgeInformation.apiVersions).withTTL(60 * 60 * 24).build();
+    }
+
+    @Override
+    public @Nullable ThingUID getThingUID(ServiceInfo serviceInfo) {
+        String ipAddress = discoverBridge(serviceInfo).shcIpAddress;
+        if (!ipAddress.isBlank()) {
+            return new ThingUID(BoschSHCBindingConstants.THING_TYPE_SHC, ipAddress.replace('.', '-'));
+        }
+        return null;
+    }
+
+    protected PublicInformation discoverBridge(ServiceInfo serviceInfo) {
+        logger.trace("Discovering serviceInfo {}", serviceInfo);
+
+        if (serviceInfo.getHostAddresses() != null && serviceInfo.getHostAddresses().length > 0
+                && !serviceInfo.getHostAddresses()[0].isEmpty()) {
+            String address = serviceInfo.getHostAddresses()[0];
+            logger.trace("Discovering InetAddress {}", address);
+            // store all information for later access
+            bridgeInformation = getPublicInformationFromPossibleBridgeAddress(address);
+        }
+
+        return bridgeInformation;
+    }
+
+    protected PublicInformation getPublicInformationFromPossibleBridgeAddress(String ipAddress) {
+
+        String url = String.format("https://%s:8446/smarthome/public/information", ipAddress);
+        logger.trace("Discovering ipAddress {}", url);
+        try {
+            httpClient.start();
+            ContentResponse contentResponse = httpClient.newRequest(url).method(HttpMethod.GET).send();
+            // check HTTP status code
+            if (!HttpStatus.getCode(contentResponse.getStatus()).isSuccess()) {
+                logger.debug("Discovering failed with status code: {}", contentResponse.getStatus());
+                return new PublicInformation();
+            }
+            // get content from response
+            String content = contentResponse.getContentAsString();
+            logger.trace("Discovered SHC - public info {}", content);
+            PublicInformation bridgeInfo = gson.fromJson(content, PublicInformation.class);
+            if (bridgeInfo.shcIpAddress != null) {
+                return bridgeInfo;
+            }
+        } catch (InterruptedException | TimeoutException | ExecutionException e) {
```

Comment:
`InterruptedException` should be handled separately because of the thread interrupt flag

Reply:
There is now a `catch` clause for all remaining Exceptions, but according to SonarLint we need an explicit catch clause for `InterruptedException` that contains `Thread.currentThread().interrupt();`, see https://rules.sonarsource.com/java/RSPEC-2142

If you don't have it yet, I can recommend the SonarLint plugin. You can simply install it via the Eclipse Marketplace.